src/Deinterleave.cpp

```diff
@@ -827,8 +1284,257 @@ void deinterleave_vector_test() {
     Expr vec_x = Variable::make(Int(32, 4), "vec_x");
     Expr vec_y = Variable::make(Int(32, 4), "vec_y");
     check(Shuffle::make({vec_x, vec_y}, {0, 4, 2, 6, 4, 2, 3, 7, 1, 2, 3, 4}),
-          Shuffle::make({vec_x, vec_y}, {0, 2, 4, 3, 1, 3}),
-          Shuffle::make({vec_x, vec_y}, {4, 6, 2, 7, 2, 4}));
+          Shuffle::make({vec_x, Shuffle::make_extract_element(vec_y, 0)},
```

Comment:
I'm don't think these outputs are better than the old ones. The contain nested shuffles and look like they might codegen to unnecessary shuffle instructions.

Reply:
1. This is a subjective judgement. Are there any p
2. It really doesn't matter how we represent the shuffles, like or not, during codegen legalization, at least for X86, all shuffles are expanded and then new shuffles constructed from the basic element flow. https://godbolt.org/z/qW7x17aKs

Reply:
On some platforms (Hexagon) we emit specific shuffle intrinsics that llvm preserves, and we're moving in the direction of handling more of instruction selection ourselves on all platforms (and finding ways to stop llvm from undoing our work).

To avoid subjective judgements, I'd really changes to be motivated by before/after assembly snippets that show this codegens better for some plausible pieces of input code. Otherwise I don't see what this complexity is buying us.

Reply:
> we're moving in the direction of handling more of instruction selection ourselves on all platforms (and finding ways to stop llvm from undoing our work).

FWIW, this seems like extremely misplaced work. Has any thought been/being put into actually benefiting everyone and working on fixing the final source of truth? :)

Reply:
Yes, we've thought about whether or not we should just be improving llvm's instruction selection, but there are a few problems with that approach. First, not all of our backends are llvm. E.g. the xtensa branch needs to do pretty involved fixed-point instruction selection and is not llvm-based. Hexagon is technically llvm, but in practice it's all opaque intrinsics that llvm just passes through. 

Second, instruction selection is practically easier to do at the Halide level because it's a more restricted IR, for which we've developed a lot of relevant analysis tools (e.g. good bounds inference machinery). It's also practically easier because this is the code we understand and have control over.

The third more philosophical issue is that llvm's main performance goal is making naively written c++ code fast. When it comes to vector instruction selection, llvm is *intentionally* undoing our work, because it correctly believes that on average, it knows more about performance than some random C++ dev. It has been doing this by destructuring intrinsics in C++, deleting them entirely in LLVM IR, and replacing them with pattern matching. This can be brittle. Sometimes part of the pattern gets lifted as a loop invariant, or gets folded with a nearby shuffle, and instruction selection goes off the rails and you get bad code. Opening bugs for each case is a game of whack-a-mole.

Halide has very different users, so we're taking a different approach. Our approach has been to steadily add intrinsics inside the IR, first portable fixed-point ones (see FindIntrinsics.cpp), and now machine-specific vector instructions (See #6884). This means we can do instruction selection as IR->IR passes, and expose increasing amounts of control over instruction selection upwards to expert users. I don't think this approach would be welcome inside llvm.